pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeRangeSegmentPruner.java

```diff
@@ -0,0 +1,348 @@
+          List<Interval> childRanges = getFilterTimeRange(child);
+          if (childRanges != null) {
+            andRanges.add(childRanges);
+          }
+        }
+        return getIntersectionSortedRanges(andRanges);
+      case OR:
+        List<List<Interval>> orRanges = new ArrayList<>();
+        for (FilterQueryTree child : filterQueryTree.getChildren()) {
+          List<Interval> childRanges = getFilterTimeRange(child);
+          if (childRanges == null) {
+            return null;
+          } else {
+            orRanges.add(childRanges);
+          }
+        }
+        return getUnionSortedRanges(orRanges);
+      case EQUALITY:
+        if (filterQueryTree.getColumn().equals(_timeColumn)) {
+          long timeStamp = Long.parseLong(filterQueryTree.getValue().get(0));
+          return Collections.singletonList(new Interval(timeStamp, timeStamp));
+        }
+        return null;
+      case RANGE:
+        if (filterQueryTree.getColumn().equals(_timeColumn)) {
+          return parseRange(filterQueryTree.getValue());
+        }
+        return null;
+      default:
+        return null;
+    }
+  }
+
+  private Interval convertRangeToMS(Interval range) {
+    long min = range.min == MIN_START_TIME ? MIN_START_TIME : _timeUnit.toMillis(range.min);
+    long max = range.max == MAX_END_TIME ? MAX_END_TIME : _timeUnit.toMillis(range.max);
+    return new Interval(min, max);
+  }
+
+  private List<Interval> getIntersectionSortedRanges(List<List<Interval>> ranges) {
+    // Requires input ranges are sorted, the return ranges will be sorted
+    return getIntersectionSortedRanges(ranges, 0, ranges.size());
+  }
+
+  private List<Interval> getIntersectionSortedRanges(List<List<Interval>> ranges, int start, int end) {
+    if (start == end) {
+      return null;
+    }
+
+    if (start + 1 == end) {
+      return ranges.get(start);
+    }
+
+    int mid = start + (end - start) / 2;
+    List<Interval> ranges1 = getIntersectionSortedRanges(ranges, start, mid);
+    List<Interval> ranges2 = getIntersectionSortedRanges(ranges, mid, end);
+    return getIntersectionOfTwoSortedRanges(ranges1, ranges2);
+  }
+
+  private List<Interval> getIntersectionOfTwoSortedRanges(List<Interval> ranges1, List<Interval> ranges2) { // sorted non-overlapping ranges
+    List<Interval> res = new ArrayList<>();
+    int i = 0, j = 0;
+    while (i < ranges1.size() && j < ranges2.size()) {
+      Interval range1 = ranges1.get(i);
+      Interval range2 = ranges2.get(j);
+      if (range1.intersects(range2)) {
+        res.add(Interval.getIntersection(range1, range2));
+      }
+      if (range1.max < range2.max) {
+        i++;
+      } else {
+        j++;
+      }
+    }
+    return res;
+  }
+
+  private List<Interval> getUnionSortedRanges(List<List<Interval>> ranges) {
+    return getUnionSortedRanges(ranges, 0, ranges.size());
+  }
+
+  private List<Interval> getUnionSortedRanges(List<List<Interval>> ranges, int start, int end) {
```

Comment:
I think List<List> is complicating the algorithm slightly. We can combine these two methods into one method that sorted merge of two List<Interval>